cms/djangoapps/contentstore/api/views/course_validation.py

```diff
@@ -0,0 +1,188 @@
+                response.update(
+                    assignments=self._assignments_validation(course)
+                )
+            if get_bool_param(request, 'grades', all_requested):
+                response.update(
+                    grades=self._grades_validation(course)
+                )
+            if get_bool_param(request, 'certificates', all_requested):
+                response.update(
+                    certificates=self._certificates_validation(course)
+                )
+            if get_bool_param(request, 'updates', all_requested):
+                response.update(
+                    updates=self._updates_validation(course, request)
+                )
+
+        return Response(response)
+
+    def _required_course_depth(self, request, all_requested):
+        if get_bool_param(request, 'assignments', all_requested):
+            return 2
+        else:
+            return 0
+
+    def _dates_validation(self, course):
+        return dict(
+            has_start_date=self._has_start_date(course),
+            has_end_date=course.end is not None,
+        )
+
+    def _assignments_validation(self, course):
+        assignments, visible_assignments = self._get_assignments(course)
+        assignments_with_dates = [a for a in visible_assignments if a.due]
+
+        num_with_dates = len(assignments_with_dates)
+        num_with_dates_after_start = (
+            len([a for a in assignments_with_dates if a.due > course.start])
+            if self._has_start_date(course)
+            else 0
+        )
+        num_with_dates_before_end = (
+            len([a for a in assignments_with_dates if a.due < course.end])
+            if course.end
+            else 0
+        )
+
+        return dict(
+            total_number=len(assignments),
+            total_visible=len(visible_assignments),
+            num_with_dates=num_with_dates,
+            num_with_dates_after_start=num_with_dates_after_start,
+            num_with_dates_before_end=num_with_dates_before_end,
+        )
+
+    def _grades_validation(self, course):
+        sum_of_weights = course.grader.sum_of_weights
+        return dict(
+            sum_of_weights=sum_of_weights,
+        )
+
+    def _certificates_validation(self, course):
+        is_activated, certificates = CertificateManager.is_activated(course)
+        return dict(
+            is_activated=is_activated,
+            has_certificate=len(certificates) > 0,
+        )
+
+    def _updates_validation(self, course, request):
+        updates_usage_key = course.id.make_usage_key('course_info', 'updates')
+        updates = get_course_updates(updates_usage_key, provided_id=None, user_id=request.user.id)
+        return dict(
+            has_update=len(updates) > 0,
+        )
+
+    def _get_assignments(self, course):
+        store = modulestore()
+        sections = [store.get_item(section_usage_key) for section_usage_key in course.children]
```

Comment:
Should this be a inside a `bulk_operation` context manager? (I'm asking because I don't know)

Reply:
Yes, it is.  We have a single `bulk_operation` context manager at the top of the request `get` method that therefore gets used by all internal methods.  The `bulk_operation` remains in action throughout the call chain within the `get` method.